File: bindings/php/Micron.php
<?php
# Copyright Quad4 2026
# SPDX-License-Identifier: 0BSD

declare(strict_types=1);

/**
 * Micron markup parser bindings over libmicron (PHP FFI).
 */
final class Micron
{
    private static ?FFI $ffi = null;

    private static function libName(): string
    {
        return match (PHP_OS_FAMILY) {
            'Darwin' => 'libmicron.dylib',
            'Windows' => 'libmicron.dll',
            default => 'libmicron.so',
        };
    }

    private static function resolveLib(): string
    {
        $env = getenv('MICRON_LIB_PATH');
        if (is_string($env) && $env !== '' && is_file($env)) {
            return $env;
        }

        $name = self::libName();
        $here = __DIR__;
        $candidates = [
            $here . '/native/' . $name,
            $here . '/../../dist/' . $name,
            getcwd() . '/dist/' . $name,
        ];
        foreach ($candidates as $path) {
            if (is_file($path)) {
                return $path;
            }
        }
        throw new RuntimeException(
            'libmicron not found; set MICRON_LIB_PATH or place the shared library under bindings/php/native or dist/'
        );
    }

    private static function ffi(): FFI
    {
        if (self::$ffi !== null) {
            return self::$ffi;
        }
        if (!extension_loaded('ffi')) {
            throw new RuntimeException('PHP FFI extension is required (enable extension=ffi and ffi.enable=true)');
        }
        $cdef = <<<'CDEF'
char *micron_convert(const char *markup, int dark_theme, int force_monospace);
char *micron_parse_header_tags(const char *markup);
char *micron_collect_form_fields(const char *inputs_json);
char *micron_build_request_payload(const char *fields_json, const char *destination, const char *fields_spec);
void micron_free(char *ptr);
CDEF;
        self::$ffi = FFI::cdef($cdef, self::resolveLib());
        return self::$ffi;
    }

    private static function takeString(?FFI\CData $ptr): string
    {
        if ($ptr === null) {
            return '';
        }
        try {
            return FFI::string($ptr);
        } finally {
            self::ffi()->micron_free($ptr);
        }
    }

    public static function convert(string $markup, bool $darkTheme = true, bool $forceMonospace = true): string
    {
        return self::takeString(
            self::ffi()->micron_convert($markup, $darkTheme ? 1 : 0, $forceMonospace ? 1 : 0)
        );
    }

    /** @return array{fg: string, bg: string} */
    public static function parseHeaderTags(string $markup): array
    {
        $data = json_decode(self::takeString(self::ffi()->micron_parse_header_tags($markup)), true);
        return [
            'fg' => (string)($data['fg'] ?? ''),
            'bg' => (string)($data['bg'] ?? ''),
        ];
    }

    /** @param list<array<string, mixed>> $inputs */
    public static function collectFormFields(array $inputs): array
    {
        $json = json_encode($inputs, JSON_THROW_ON_ERROR);
        $data = json_decode(self::takeString(self::ffi()->micron_collect_form_fields($json)), true);
        return is_array($data) ? $data : [];
    }

    /** @param array<string, string> $fields */
    public static function buildRequestPayload(array $fields, string $destination, string $fieldsSpec): array
    {
        $json = json_encode($fields, JSON_THROW_ON_ERROR);
        $data = json_decode(
            self::takeString(
                self::ffi()->micron_build_request_payload($json, $destination, $fieldsSpec)
            ),
            true
        );
        return is_array($data) ? $data : [];
    }
}

File: bindings/ruby/lib/micron.rb
# frozen_string_literal: true
# Copyright Quad4 2026
# SPDX-License-Identifier: 0BSD

require "ffi"
require "json"

module Micron
  extend FFI::Library

  def self.lib_name
    case FFI::Platform::OS
    when "darwin" then "libmicron.dylib"
    when "windows" then "libmicron.dll"
    else "libmicron.so"
    end
  end

  def self.resolve_lib
    env = ENV["MICRON_LIB_PATH"]
    return env if env && File.file?(env)

    name = lib_name
    candidates = [
      File.expand_path("../native/#{name}", __dir__),
      File.expand_path("native/#{name}", __dir__),
      File.expand_path("../../../dist/#{name}", __dir__),
      File.expand_path("dist/#{name}"),
    ]
    candidates.find { |p| File.file?(p) } || name
  end

  ffi_lib resolve_lib

  attach_function :micron_convert, [:string, :int, :int], :pointer
  attach_function :micron_parse_header_tags, [:string], :pointer
  attach_function :micron_collect_form_fields, [:string], :pointer
  attach_function :micron_build_request_payload, [:string, :string, :string], :pointer
  attach_function :micron_free, [:pointer], :void

  module_function

  def take_string(ptr)
    return "" if ptr.null?

    begin
      ptr.read_string
    ensure
      micron_free(ptr)
    end
  end

  def convert(markup, dark_theme: true, force_monospace: true)
    take_string(micron_convert(markup.to_s, dark_theme ? 1 : 0, force_monospace ? 1 : 0))
  end

  def parse_header_tags(markup)
    JSON.parse(take_string(micron_parse_header_tags(markup.to_s)))
  end

  def collect_form_fields(inputs)
    JSON.parse(take_string(micron_collect_form_fields(JSON.generate(inputs))))
  end

  def build_request_payload(fields, destination, fields_spec)
    JSON.parse(
      take_string(
        micron_build_request_payload(JSON.generate(fields), destination.to_s, fields_spec.to_s)
      )
    )
  end
end
